### PR TITLE
[SPARK-35735][SQL]  Take into account day-time interval fields in cast

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Support take into account day-time interval field in cast.

### Why are the changes needed?
To conform to the SQL standard.

### Does this PR introduce _any_ user-facing change?
An user can use `cast(str, DayTimeInterval(DAY, HOUR))`, for instance.

### How was this patch tested?
Added UT.

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
Ping @MaxGekk

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala [hunk 1]
```diff
@@ -997,12 +997,12 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
...
-      Seq("INTERVAL '-106751991 04:00:54.775809' YEAR TO MONTH",
-        "INTERVAL '106751991 04:00:54.775808' YEAR TO MONTH").foreach { interval =>
-        val e = intercept[IllegalArgumentException] {
+      Seq("INTERVAL '-106751991 04:00:54.775809' DAY TO SECOND",
+        "INTERVAL '106751991 04:00:54.775808' DAY TO SECOND").foreach { interval =>
+        val e = intercept[ArithmeticException] {
           cast(Literal.create(interval), DayTimeIntervalType()).eval()
         }.getMessage
-        assert(e.contains("Interval string must match day-time format of"))
+        assert(e.contains("long overflow"))
```
Comment (AngersZhuuuu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:1005]: Here is a mistake when handle SPARK-35112

**Comment by AngersZhuuuu** [comment]
ping @MaxGekk @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 2]
```diff
@@ -213,24 +222,69 @@ object IntervalUtils {
...
+    (s"(?i)^INTERVAL\\s+([+|-])?\\'$dayMinutePatternString\\'\\s+DAY\\s+TO\\s+MINUTE$$").r
+
   private val daySecondPatternString =
-    "([+|-])?(\\d+) (\\d{1,2}):(\\d{1,2}):(\\d{1,2})(\\.\\d{1,9})?"
+    s"([+|-])?$dayPattern $normalPattern:$normalPattern:$normalPattern$microPattern?"
   private val daySecondRegex = (s"^$daySecondPatternString$$").r
   private val daySecondLiteralRegex =
     (s"(?i)^INTERVAL\\s+([+|-])?\\'$daySecondPatternString\\'\\s+DAY\\s+TO\\s+SECOND$$").r
 
+  private val hourMinutePatternString = s"([+|-])?$normalPattern:$normalPattern"
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:252]: The hour field at the first position is not restricted by 2 digits.

**Review by MaxGekk — CHANGES_REQUESTED** [review]
Please, avoid the restriction by 2 digits for fields at the `start` position. Also cc @sarutak @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 3]
```diff
@@ -213,24 +222,69 @@ object IntervalUtils {
...
   private def toYMInterval(yearStr: String, monthStr: String, sign: Int): Int = {
-    safeToYMInterval {
+    safeToInterval {
       val years = toLongWithRange(YEAR, yearStr, 0, Integer.MAX_VALUE / MONTHS_PER_YEAR)
       val totalMonths = sign * (years * MONTHS_PER_YEAR + toLongWithRange(MONTH, monthStr, 0, 11))
       Math.toIntExact(totalMonths)
     }
   }
 
+  private val dayPattern = "(\\d+)"
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: > The upper bound to fit to `Long` is 106751991, so, let's restrict the number of digits by `{1,9}`. Do you have a check for that?

Check all first unit and update UT.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: > The hour field at the first position is not restricted by 2 digits.

Yea, check this and add UT

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 4]
```diff
@@ -119,18 +119,27 @@ object IntervalUtils {
...
+    (DT.DAY, DT.DAY) -> Seq("[+|-]d", "INTERVAL [+|-]'[+|-]d' DAY"),
+    (DT.DAY, DT.HOUR) -> Seq("[+|-]d h", "INTERVAL [+|-]'[+|-]d h' DAY TO HOUR"),
+    (DT.DAY, DT.MINUTE) -> Seq("[+|-]d h:m", "INTERVAL [+|-]'[+|-]d h:m' DAY TO MINUTE"),
+    (DT.DAY, DT.SECOND) -> Seq("[+|-]d h:m:s.n", "INTERVAL [+|-]'[+|-]d h:m:s.n' DAY TO SECOND"),
+    (DT.HOUR, DT.HOUR) -> Seq("[+|-]h", "INTERVAL [+|-]'[+|-]h' HOUR TO HOUR"),
+    (DT.HOUR, DT.MINUTE) -> Seq("[+|-]h:m", "INTERVAL [+|-]'[+|-]h:m' HOUR TO MINUTE"),
+    (DT.HOUR, DT.SECOND) -> Seq("[+|-]h:m:s.n", "INTERVAL [+|-]'[+|-]h:m:s.n' HOUR TO SECOND"),
+    (DT.MINUTE, DT.MINUTE) -> Seq("[+|-]m", "INTERVAL [+|-]'[+|-]m' MINUTE TO MINUTE"),
+    (DT.MINUTE, DT.SECOND) -> Seq("[+|-]m:s.n", "INTERVAL [+|-]'[+|-]m:s.n' MINUTE TO SECOND"),
+    (DT.SECOND, DT.SECOND) -> Seq("[+|-]s.n", "INTERVAL [+|-]'[+|-]s.n' SECOND TO SECOND")
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: Done

**Comment by AngersZhuuuu** [comment]
ping @MaxGekk  Can help to re-trigger the jenkins test?

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
Hmm, Jenkins seems down today.

**Comment by MaxGekk** [comment]
@AngersZhuuuu You can re-trigger yourself by making an empty commit:
```
git commit --allow-empty -m "Trigger build"
```

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu You can re-trigger yourself by making an empty commit:
> 
> ```
> git commit --allow-empty -m "Trigger build"
> ```

Can work now, any more suggestion?

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @AngersZhuuuu and @sarutak for review.

**Pull request closed without merging**